### DISCUSSION TITLE
Custom datasource: get digests from md5sums, sha256sums files

### DISCUSSION THREAD

**Discussion by amezin** [renovatebot/renovate#26022] [Suggest an Idea]
### Tell us more.

For example, look at Ubuntu images: https://releases.ubuntu.com/23.10.1/

There is a https://releases.ubuntu.com/23.10.1/SHA256SUMS file, and checksums from that file can be used to verify the downloaded files (for example, in CMake's ExternalProject - https://cmake.org/cmake/help/latest/module/ExternalProject.html#examples).

Other examples:
- Qt source code tarballs: https://download.qt.io/archive/qt/6.6/6.6.1/single/md5sums.txt
- POCO: https://pocoproject.org/releases/poco-1.12.5p1/SHASUMS256.txt
- Fedora images: https://download.fedoraproject.org/pub/fedora/linux/releases/39/Workstation/x86_64/iso/Fedora-Workstation-39-1.5-x86_64-CHECKSUM
- Debian: https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/SHA512SUMS
- ...

Custom datasource with `html` format + regex manager can update versions for such downloads, but can't currently update the checksum.

It can be solved if custom datasource allowed returning `checksumUrl` and `downloadUrl` and, when digest is requested, downloaded `checksumUrl` and extracted the checksum from that file (using `downloadUrl` to find the correct row).

Unfortunately, it's not possible to use the checksum file as the datasource - because usually the file only contains data about one version/release.

**Comment by secustor** [comment]
For this to implement you basically need to add a `getDigest` method to `CustomDatasource`, tough this requires changes on the interface too.

**Reply by amezin** [reply]
So, this behavior is acceptable, right? I want to be sure before I waste time on the implementation.

**Reply by secustor** [reply]
@rarkins @viceice WDYT?

**Reply by amezin** [reply]
> tough this requires changes on the interface too

You mean [schema](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/custom/schema.ts) changes?

Both `checksumUrl` and `downloadUrl` are parts of [`Release` interface](https://github.com/renovatebot/renovate/blob/a065776a6ce54f5d4c5ade94ff4e536e543e5380/lib/modules/datasource/types.ts#L50) already - but I can't find where they are consumed

**Reply by viceice** [reply]
general idea is ok, but implementation can be challenging/ very difficult

**Reply by amezin** [reply]
So this requires either:

1) `getDigest` signature to be changed - to pass `downloadUrl` and `checksumUrl` into the function (probably some subset of `ReleaseResult` should be passed as the second argument, instead of just `newValue` string)

or

2) `DigestConfig` and `GetDigestInputConfig` should include `downloadUrl` and `checksumUrl` (requires less changes, in terms of lines of code)

which is the correct way?

**Reply by amezin** [reply]
Or, maybe, this shouldn't even be part of custom datasource, and should be implemented in custom (regex) manager instead? After all, checksums are required because of the file format. I may get the version from `endoflife-date`, but still want to update a checksum (for an archive downloaded by Makefile or CMake-based build system, for example).

**Reply by ivankatliarchuk** [reply]
Any plans to implement it or create proof of concept?

**Reply by secustor** [reply]
https://github.com/renovatebot/renovate/pull/26299 will add the ability to get digests from the custom datasource, but the values have to come from the same API call. 

From the maintainers perspective first a consensus must be reached here and then this discussion will be transferred to an issue. If that's the case we signal that we would accept contributions to that feature. 
There's always to possibility to simply provide a PR, tough major rewrites could be requested until it reaches a mergeable state.

**Reply by ivankatliarchuk** [reply]
Hard to disagree. The pull request itself is not an issue, is more about what the desired approach is. I have a case when value is on S3 bucket and dependency is in Gitlab. example https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.25/index.html. Release https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/releases

**Comment by amezin** [comment]
By the way, similar checksum extraction is implemented in hashicorp's go-getter: https://github.com/hashicorp/go-getter/blob/main/checksum.go

**Comment by btquanto** [comment]
I was looking for the solution for this. I'm using a custom artifactory and applying Renovate for a bash script. The code requires to do a checksum for the downloaded artifact. The problem is that the artifactory API only returns the versions list, and there's another API that receives the version number and returns the meta information, including the SHA256 checksum.

What I ended up doing is to give up managing the SHA256 digest with Renovate, and instead added a piece of code to call the API, extract the checksum and validate it at run time.
